app/routes/datasets/datasets.py

```diff
@@ -14,11 +17,33 @@
     "",
     response_class=ORJSONResponse,
     tags=["Datasets"],
-    response_model=DatasetsResponse,
+    response_model=Union[PaginatedDatasetsResponse, DatasetsResponse],
 )
-async def get_datasets() -> DatasetsResponse:
+async def get_datasets(
+    request: Request,
+    page_number: Optional[int] = Query(
+        default=None, alias="page[number]", ge=1, description="The page number."
+    ),
+    page_size: Optional[int] = Query(
+        default=None,
+        alias="page[size]",
+        ge=1,
+        description="The number of datasets per page. Default is `10`.",
+    ),
+) -> Union[PaginatedDatasetsResponse, DatasetsResponse]:
     """Get list of all datasets."""
+    try:
+        data, links, meta = await paginate_collection(
```

Comment:
This isn't a huge issue, but this does feel a little reversed from what I'd expect. You always call the paginate function, check in the paginate function if the condition pagination are met (size or page exists), and then check if the pagination function was able to generate links or metadata to see what type of response to return.

I think it might be clearer how we decide the response method if instead the method checked if size or page exists first, and decides the response based on that. 

Also might be good to include that in the documentation now for the endpoint, e.g. "Will attempt to paginate if size or page is provided, or will return all documents as a response." In fact, that'd be good enough for me, I think I just this way made me do some digging to remember how we choose the response type.

Reply:
Thank you for this, @jterry64. I thought about removing the check from `paginate_collection` near the end of the implementation and I just didn't do it. I'm going to do that refactoring. It'll clean things up.

Early on, I wanted to have `get_datasets` (path operation function) to make as few decisions as possible. But, as you can see, I still wound up making decisions there. And it is more challenging to follow. 😂

Reply:
Docs updated here: [e1200d0](https://github.com/wri/gfw-data-api/pull/327/commits/e1200d0a6b1989b146f8dea5c831f552736bf3a4)

Reply:
Also, I was excited that none of the unit tests had to change, except for getting rid of the unneeded legacy behavior. As a result, I could rapidly move things around because the tests did not constrain the internal implementation. They are also fast to execute since they don't drag in the DB.